File: app/src/androidTest/java/io/github/fvasco/pinpoi/importer/ImporterFacadeTest.java
package io.github.fvasco.pinpoi.importer;

import android.content.Context;
import android.test.RenamingDelegatingContext;

import org.junit.Test;

import java.io.IOException;
import java.net.URL;

import io.github.fvasco.pinpoi.dao.PlacemarkCollectionDao;
import io.github.fvasco.pinpoi.dao.PlacemarkDao;
import io.github.fvasco.pinpoi.model.PlacemarkCollection;

/**
 * @author Francesco Vasco
 */
public class ImporterFacadeTest extends AbstractImporterTestCase {

    private PlacemarkCollectionDao placemarkCollectionDao;
    private PlacemarkDao placemarkDao;
    private Context context;

    @Override
    protected void setUp() throws Exception {
        super.setUp();
        context = new RenamingDelegatingContext(getContext(), "test_");
        placemarkCollectionDao = new PlacemarkCollectionDao(context);
        placemarkDao = new PlacemarkDao(context);
        placemarkCollectionDao.open();
        placemarkDao.open();
    }

    @Override
    protected void tearDown() throws Exception {
        super.tearDown();
        placemarkDao.close();
        placemarkCollectionDao.close();
    }

    private PlacemarkCollection insertPlacemarkCollection(String resource) {
        final PlacemarkCollection pc = new PlacemarkCollection();
        pc.setName(resource);
        final URL resourceUrl = getClass().getResource(resource);
        pc.setSource(resourceUrl == null ? resource : resourceUrl.toString());
        placemarkCollectionDao.insert(pc);
        return pc;
    }

    @Test
    public void testImportPlacemarksKml() throws Exception {
        final PlacemarkCollection pc = insertPlacemarkCollection("test2.kml");
        final ImporterFacade importerFacade = new ImporterFacade(context);
        int count = importerFacade.importPlacemarks(pc);
        assertEquals(2, count);

        assertEquals(count, pc.getPoiCount());
        final long lastUpdate = pc.getLastUpdate();
        assertTrue(lastUpdate > 0);

        assertEquals(count, placemarkDao.findAllPlacemarkByCollectionId(pc.getId()).size());
    }

    @Test
    public void testImportPlacemarksWrong() throws Exception {
        final PlacemarkCollection pc = insertPlacemarkCollection("wrong");
        pc.setLastUpdate(1);
        pc.setPoiCount(10);
        final ImporterFacade importerFacade = new ImporterFacade(context);
        try {
            importerFacade.importPlacemarks(pc);
            fail();
        } catch (final IOException e) {
            // ok
        }
        assertEquals(1, pc.getLastUpdate());
        assertEquals(10, pc.getPoiCount());
    }

    @Test
    public void testImportPlacemarksArc() throws Exception {
        final PlacemarkCollection pc = insertPlacemarkCollection("asc.txt");
        final ImporterFacade importerFacade = new ImporterFacade(context);
        int count = importerFacade.importPlacemarks(pc);
        assertEquals(2, count);
        assertEquals(count, pc.getPoiCount());
        assertEquals(count, placemarkDao.findAllPlacemarkByCollectionId(pc.getId()).size());
    }
}

File: app/src/androidTest/java/io/github/fvasco/pinpoi/importer/ImporterFacadeTest.kt
package io.github.fvasco.pinpoi.importer

import android.support.test.runner.AndroidJUnit4
import io.github.fvasco.pinpoi.dao.PlacemarkCollectionDao
import io.github.fvasco.pinpoi.dao.PlacemarkDao
import io.github.fvasco.pinpoi.model.PlacemarkCollection
import org.junit.After
import org.junit.Assert.*
import org.junit.Before
import org.junit.Test
import org.junit.runner.RunWith
import java.io.IOException

/**
 * @author Francesco Vasco
 */
@RunWith(AndroidJUnit4::class)
class ImporterFacadeTest : AbstractImporterTestCase() {

    private lateinit var placemarkCollectionDao: PlacemarkCollectionDao
    private lateinit var placemarkDao: PlacemarkDao

    @Before
    fun setUp() {
        placemarkCollectionDao = PlacemarkCollectionDao(context)
        placemarkDao = PlacemarkDao(context)
        placemarkCollectionDao.open()
        placemarkDao.open()
    }

    @After
    fun tearDown() {
        placemarkDao.close()
        placemarkCollectionDao.close()
    }

    private fun insertPlacemarkCollection(resource: String): PlacemarkCollection {
        val pc = PlacemarkCollection()
        pc.name = resource
        val resourceUrl = javaClass.getResource(resource)
        pc.source = if (resourceUrl == null) resource else resourceUrl.toString()
        placemarkCollectionDao.insert(pc)
        return pc
    }

    @Test
    fun testImportPlacemarksKml() {
        val pc = insertPlacemarkCollection("test2.kml")
        val importerFacade = ImporterFacade(context)
        val count = importerFacade.importPlacemarks(pc)
        assertEquals(2, count)

        assertEquals(count, pc.poiCount)
        val lastUpdate = pc.lastUpdate
        assertTrue(lastUpdate > 0)

        assertEquals(count, placemarkDao.findAllPlacemarkByCollectionId(pc.id).size)
    }

    @Test
    fun testImportPlacemarksWrong() {
        val pc = insertPlacemarkCollection("wrong")
        pc.lastUpdate = 1
        pc.poiCount = 10
        val importerFacade = ImporterFacade(context)
        try {
            importerFacade.importPlacemarks(pc)
            fail()
        } catch (e: IOException) {
            // ok
        }

        assertEquals(1, pc.lastUpdate)
        assertEquals(10, pc.poiCount)
    }

    @Test
    fun testImportPlacemarksAsc() {
        val pc = insertPlacemarkCollection("asc.txt")
        val importerFacade = ImporterFacade(context)
        val count = importerFacade.importPlacemarks(pc)
        assertEquals(3, count)
        assertEquals(count, pc.poiCount)
        assertEquals(count, placemarkDao.findAllPlacemarkByCollectionId(pc.id).size)
    }

    @Test
    fun testImportPlacemarksCsvLanLon() {
        val pc = insertPlacemarkCollection("csv.txt")
        val importerFacade = ImporterFacade(context)
        val count = importerFacade.importPlacemarks(pc)
        importerFacade.fileFormatFilter = FileFormatFilter.CSV_LAT_LON
        assertEquals(3, count)
        assertEquals(count, pc.poiCount)
        assertEquals(count, placemarkDao.findAllPlacemarkByCollectionId(pc.id).size)
    }

    @Test
    fun testImportPlacemarksCsvLonLat() {
        val pc = insertPlacemarkCollection("csv.txt")
        val importerFacade = ImporterFacade(context)
        importerFacade.fileFormatFilter = FileFormatFilter.CSV_LON_LAT
        val count = importerFacade.importPlacemarks(pc)
        assertEquals(4, count)
        assertEquals(count, pc.poiCount)
        assertEquals(count, placemarkDao.findAllPlacemarkByCollectionId(pc.id).size)
    }
}